src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+    std::map<AccountID, std::vector<LedgerStateEntry>> offersByAccount;
+    for (auto const& kv : offers)
+    {
+        auto const& key = kv.first;
+        auto const& sellerID = key.offer().sellerID;
+        offersByAccount[sellerID].emplace_back(load(self, key));
+    }
+    return offersByAccount;
+}
+
+LedgerStateEntry
+LedgerState::loadBestOffer(Asset const& buying, Asset const& selling)
+{
+    return mImpl->loadBestOffer(*this, buying, selling);
+}
+
+LedgerStateEntry
+LedgerState::Impl::loadBestOffer(LedgerState& self, Asset const& buying,
+                                 Asset const& selling)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    auto le = getBestOffer(buying, selling, {});
+    return le ? load(self, LedgerEntryKey(*le)) : LedgerStateEntry();
+}
+
+LedgerStateHeader
+LedgerState::loadHeader()
+{
+    return mImpl->loadHeader(*this);
+}
+
+LedgerStateHeader
+LedgerState::Impl::loadHeader(LedgerState& self)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActiveHeader)
+    {
+        throw std::runtime_error("LedgerStateHeader is active");
+    }
+
+    mActiveHeader = std::make_shared<HeaderImpl>(self, mHeader);
+    return LedgerStateHeader(mActiveHeader);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::loadOffersByAccountAndAsset(AccountID const& accountID,
+                                         Asset const& asset)
+{
+    return mImpl->loadOffersByAccountAndAsset(*this, accountID, asset);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::Impl::loadOffersByAccountAndAsset(LedgerState& self,
+                                               AccountID const& accountID,
+                                               Asset const& asset)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    std::vector<LedgerStateEntry> res;
+    auto offers = getOffersByAccountAndAsset(accountID, asset);
+    for (auto const& kv : offers)
+    {
+        auto const& key = kv.first;
+        res.emplace_back(load(self, key));
+    }
+    return res;
+}
+
+ConstLedgerStateEntry
+LedgerState::loadWithoutRecord(LedgerKey const& key)
+{
+    return mImpl->loadWithoutRecord(*this, key);
+}
+
+ConstLedgerStateEntry
+LedgerState::Impl::loadWithoutRecord(LedgerState& self, LedgerKey const& key)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActive.find(key) != mActive.end())
+    {
+        throw std::runtime_error("Key already loaded");
+    }
+
+    auto newest = getNewestVersion(self, key);
+    if (!newest.entry)
+    {
+        return {};
+    }
+
+    auto impl =
+        std::make_shared<ConstLedgerStateEntryImpl>(self, *newest.entry);
+    mActive.emplace(key, impl);
+    return ConstLedgerStateEntry(impl);
+}
+
+void
+LedgerState::rollback()
+{
+    mImpl->rollback(getIdentifier());
+    mImpl.reset();
+}
+
+void
+LedgerState::Impl::rollback(Identifier id)
+{
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    mActive.clear();
+    mActiveHeader.reset();
+
+    mParent.rollbackChild(id);
```

Comment:
this looks a bit weird
State in its own rollback initiates parent->rollbackChild, and its name suggest that it will do rollback on this State object, which parent does not do. What about renaming rollbackChild to clearChild or resetChild?

Reply:
I agree the naming is a little strange, but I chose this name since `rollbackChild` is to `rollback` as `commitChild` is to `commit`. I'm not sure if `clearChild` or `resetChild` are much better since they also imply that they are doing something to the child, which they are not.